### PR TITLE
Support standard-json based compilation

### PR DESCRIPTION
**Pull request by pipermerriam**
Continuation of https://github.com/pipermerriam/populus/pull/294

### What was wrong?

Solidity `>=0.4.9` uses a new compilation standard using the `--standard-json` flag.  Populus doesn't know how to handle this compiler output.

### How was it fixed?

WIP. 

#### Cute Animal Picture

cute baby monkeys 6

### CODE REVIEW

**Comment by pipermerriam** [comment]
@cag thanks for starting this.

File: populus/contracts/backends/project.py [hunk 1]
```diff
@@ -24,10 +20,4 @@ def get_contract_identifier(self, contract_name):
         return contract_name
 
     def get_all_contract_data(self):
-        project_contract_data = {
-            contract_name: contract_data
-            for contract_name, contract_data
-            in self.chain.project.compiled_contract_data.items()
-            if is_project_contract(self.chain.project.contracts_source_dir, contract_data)
-        }
-        return project_contract_data
```
Comment (pipermerriam) [inline populus/contracts/backends/project.py]: This change should be reverted.

File: populus/contracts/contract.py [hunk 2]
```diff
@@ -35,7 +35,9 @@ def build_populus_meta(chain, contract_identifier, contract_data):
         'is_test_contract',
         is_test_contract(chain.project.tests_dir, contract_data),
     )
-    yield 'contract_type_name', normalize_class_name(contract_identifier)
+    if isinstance(contract_identifier, tuple):
```
Comment (pipermerriam) [inline populus/contracts/contract.py]: this change should be reverted.

File: populus/contracts/provider.py [hunk 3]
```diff
@@ -83,26 +77,10 @@ def is_contract_available(self, contract_identifier):
...
 
     def are_contract_dependencies_available(self, contract_identifier):
-        full_dependency_graph = get_shallow_dependency_graph(
-            self.chain.project.compiled_contract_data,
-        )
-        contract_dependencies = get_recursive_contract_dependencies(
-            contract_identifier,
-            full_dependency_graph,
-        )
+        cdata = self.get_contract_data(contract_identifier)
```
Comment (pipermerriam) [inline populus/contracts/provider.py]: rename abbreviation to `contract_data` for consistency

File: populus/contracts/provider.py [hunk 4]
```diff
@@ -130,21 +108,10 @@ def deploy_contract(self,
...
         the provided deployment arguments
         """
-        full_dependency_graph = get_shallow_dependency_graph(
-            self.chain.project.compiled_contract_data,
-        )
-        contract_dependencies = get_recursive_contract_dependencies(
-            contract_identifier,
-            full_dependency_graph,
-        )
+        cdata = self.get_contract_data(contract_identifier)
```
Comment (pipermerriam) [inline populus/contracts/provider.py]: rename to `contract_data` for consistency

File: populus/contracts/provider.py [hunk 5]
```diff
@@ -193,7 +160,7 @@ def get_contract_data(self, contract_identifier):
         Returns a dictionary containing the compiler output for the given
         contract identifier.
         """
-        for backend in self.provider_backends.value():
+        for backend in self.provider_backends.values():
```
Comment (pipermerriam) [inline populus/contracts/provider.py]: Get coverage for this line of code as it is clearly not being hit by CI

File: populus/contracts/provider.py [hunk 6]
```diff
@@ -260,14 +229,22 @@ def _link_bytecode(self, bytecode):
...
-            (link_reference, self.chain.provider.get_contract(link_reference.full_name).address)
-            for link_reference
-            in find_link_references(
-                bytecode,
-                self.get_all_contract_names(),
+        bytecode_key = 'bytecode'
+        linkrefs_key = 'linkrefs'
+        if is_runtime:
+            bytecode_key += '_runtime'
+            linkrefs_key += '_runtime'
```
Comment (pipermerriam) [inline populus/contracts/provider.py]: This is gross....  how to do it better though?...

File: populus/contracts/provider.py [hunk 7]
```diff
@@ -260,14 +229,22 @@ def _link_bytecode(self, bytecode):
...
+
+        resolved_link_references = tuple(
+            (
+                link_reference,
+                # TODO: need to convert to whatever the new contract key format is
```
Comment (pipermerriam) [inline populus/contracts/provider.py]: Remove comment

File: populus/plugin.py [hunk 8]
```diff
@@ -19,7 +19,7 @@ def project(request):
     project = Project()
 
     project.fill_contracts_cache(contracts, mtime)
-    request.config.cache.set(CACHE_KEY_CONTRACTS, project.compiled_contract_data)
+    request.config.cache.set(CACHE_KEY_CONTRACTS, project.compiled_contract_data.to_dict())
```
Comment (pipermerriam) [inline populus/plugin.py]: This line will throw an error.  Get coverage for it.

File: populus/utils/linking.py [hunk 9]
```diff
@@ -109,28 +155,16 @@ def link_bytecode(bytecode, link_reference_values):
     Given the bytecode for a contract, and it's dependencies in the form of
     {contract_name: address} this functino returns the bytecode with all of the
     link references replaced with the dependency addresses.
+
+    TODO: validate that the provided values are of the appropriate length
```
Comment (pipermerriam) [inline populus/utils/linking.py]: Create an "Easy Pickings" ticket for this.

**Comment by pipermerriam** [comment]
TODO

* Add a test for long file paths when using the combined-json backend that blows up if the contract-name was truncated.
* For `solc>=0.4.9` test both combined and standard json.
* For `solc<0.4.9` throw an error in the standard-json compiler backend.

**Comment by cag** [comment]
No problem! Sorry I've been scarce in following up. It's been very busy for
me..

**Comment by pipermerriam** [comment]
Nothing to apologize for.  You got me started which was a big help.  Hoping to get this released tomorrow.

**Pull request merged**